pkg/provisioner/ironic/ironic.go

```diff
@@ -1,62 +1,313 @@
 package ironic
 
 import (
+	"fmt"
+	"strings"
 	"time"
 
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/noauth"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/nodes"
+	"github.com/gophercloud/gophercloud/openstack/baremetal/v1/ports"
+
 	"github.com/pkg/errors"
 
 	"github.com/go-logr/logr"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 
 	metalkubev1alpha1 "github.com/metalkube/baremetal-operator/pkg/apis/metalkube/v1alpha1"
+	"github.com/metalkube/baremetal-operator/pkg/bmc"
 	"github.com/metalkube/baremetal-operator/pkg/provisioner"
 )
 
 var log = logf.Log.WithName("ironic")
 var deprovisionRequeueDelay = time.Second * 10
 
 // Ironic-specific provisioner factory
-type provisionerFactory struct{}
+type provisionerFactory struct {
+	// Where is ironic?
+	ironicEndpoint string
+	// The image to deploy on new hosts
+	instanceImageSource string
+	// The checksum for the instanceImageSource
+	instanceImageChecksum string
+}
 
 // New returns a new Ironic ProvisionerFactory
-func NewFactory() provisioner.ProvisionerFactory {
-	return &provisionerFactory{}
+func NewFactory(ironicEndpoint, instanceImageSource, instanceImageChecksum string) provisioner.ProvisionerFactory {
+	return &provisionerFactory{
+		ironicEndpoint:        ironicEndpoint,
+		instanceImageSource:   instanceImageSource,
+		instanceImageChecksum: instanceImageChecksum,
+	}
 }
 
 // Provisioner implements the provisioning.Provisioner interface
 // and uses Ironic to manage the host.
 type ironicProvisioner struct {
 	// the host to be managed by this provisioner
 	host *metalkubev1alpha1.BareMetalHost
+	// a shorter path to the provisioning status data structure
+	status *metalkubev1alpha1.ProvisionStatus
+	// access parameters for the BMC
+	bmcAccess *bmc.AccessDetails
+	// credentials to log in to the BMC
+	bmcCreds bmc.Credentials
+	// a client for talking to ironic
+	client *gophercloud.ServiceClient
 	// a logger configured for this host
 	log logr.Logger
+	// The image to deploy on new hosts
+	instanceImageSource string
+	// The checksum for the instanceImageSource
+	instanceImageChecksum string
 }
 
 // New returns a new Ironic Provisioner
-func (f *provisionerFactory) New(host *metalkubev1alpha1.BareMetalHost) (provisioner.Provisioner, error) {
+func (f *provisionerFactory) New(host *metalkubev1alpha1.BareMetalHost, bmcCreds bmc.Credentials) (provisioner.Provisioner, error) {
+	client, err := noauth.NewBareMetalNoAuth(noauth.EndpointOpts{
+		IronicEndpoint: f.ironicEndpoint,
+	})
+	if err != nil {
+		return nil, err
+	}
+	bmcAccess, err := bmc.NewAccessDetails(host.Spec.BMC.Address)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse BMC address information")
+	}
+	// Ensure we have a microversion high enough to get the features
+	// we need.
+	client.Microversion = "1.50"
```

Comment:
1.50 is pretty random. Use 1.56 which is very likely to be Stein final? We will need it for configdrive building through Ironic (if we plan to use it).

Reply:
I made a note in the ticket about updating to stein to include updating the microversion.